glean-core/ffi/src/event.rs

```diff
@@ -0,0 +1,127 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::iter::Iterator;
+use std::os::raw::c_char;
+
+use ffi_support::{define_handle_map_deleter, ConcurrentHandleMap, FfiStr};
+use lazy_static::lazy_static;
+
+use glean_core::metrics::{EventMetric, MetricType};
+use glean_core::{CommonMetricData, Lifetime};
+
+use crate::handlemap_ext::HandleMapExtension;
+use crate::{from_raw_string_array, RawIntArray, RawStringArray, GLEAN};
+
+lazy_static! {
+    pub static ref EVENT_METRICS: ConcurrentHandleMap<EventMetric> = ConcurrentHandleMap::new();
+}
+define_handle_map_deleter!(EVENT_METRICS, glean_destroy_event_metric);
+
+#[no_mangle]
+pub extern "C" fn glean_new_event_metric(
+    category: FfiStr,
+    name: FfiStr,
+    send_in_pings: RawStringArray,
+    send_in_pings_len: i32,
+    lifetime: i32,
+    disabled: u8,
+    extra_keys: RawStringArray,
+    extra_keys_len: i32,
+) -> u64 {
+    EVENT_METRICS.insert_with_log(|| {
+        let send_in_pings = unsafe { from_raw_string_array(send_in_pings, send_in_pings_len) };
+        let lifetime = Lifetime::try_from(lifetime)?;
+        let extra_keys = unsafe { from_raw_string_array(extra_keys, extra_keys_len) };
+
+        Ok(EventMetric::new(
+            CommonMetricData {
+                name: name.into_string(),
+                category: category.into_string(),
+                send_in_pings,
+                lifetime,
+                disabled: disabled != 0,
+            },
+            extra_keys,
+        ))
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn glean_event_should_record(glean_handle: u64, metric_id: u64) -> u8 {
+    GLEAN.call_infallible(glean_handle, |glean| {
+        EVENT_METRICS.call_infallible(metric_id, |metric| metric.should_record(&glean))
+    })
+}
+
+/// Create a HashMap<usize, String> from a pair of C arrays.
+unsafe fn from_raw_int_string_arrays(
```

Comment:
Can we move this into `ffi/src/lib.rs`? We already have `from_raw_string_array` there and it would be a good idea to keep these utility functions close to each other.

Reply:
The name of the function is already not conveying what it does.

Reply:
I've moved and renamed this -- hopefully a little better now.